click/types.py

```diff
@@ -152,29 +156,34 @@ def get_missing_message(self, param):
         return 'Choose from:\n\t%s.' % ',\n\t'.join(self.choices)
 
     def convert(self, value, param, ctx):
-        # Exact match
-        if value in self.choices:
-            return value
-
         # Match through normalization and case sensitivity
         # first do token_normalize_func, then lowercase
         # preserve original `value` to produce an accurate message in
         # `self.fail`
         normed_value = value
-        normed_choices = self.choices
+        normed_choices = {choice: choice for choice in self.choices}
 
-        if ctx is not None and \
-           ctx.token_normalize_func is not None:
+        if ctx is not None and ctx.token_normalize_func is not None:
             normed_value = ctx.token_normalize_func(value)
-            normed_choices = [ctx.token_normalize_func(choice) for choice in
-                              self.choices]
+            normed_choices = {
+                ctx.token_normalize_func(normed_choice): original
+                for normed_choice, original in normed_choices.items()
+            }
 
         if not self.case_sensitive:
-            normed_value = normed_value.lower()
-            normed_choices = [choice.lower() for choice in normed_choices]
+            if PY2:
+                lower = str.lower
```

Comment:
Using casefold below seems correct.
Should this actually be
```
if PY2:
    casefold = lamba x: x.upper().lower()
else:
    casefold = str.casefold
```

instead? `.upper().lower()` is kind of a "poor man's casefold" which will work for most things.

Reply:
I can't say I really know my way around this.  What's an example of a thing that `.upper().lower()` covers in py2 that `.lower()` doesn't?  Since I don't know I don't care either way but figure I may as well learn.  `:]`

Reply:
Sorry, I've misled you here. I'm not as clever as I think I am 😉 .

A case for casefold that I was thinking of is `ß`.
Uppercase `ß` should be `SS` -- I mean that in terms of how human languages, e.g. German, work -- so that `ß` and `ss` and `SS` all are treated as equivalent. But it seems that under py2, that's not how `upper()` on a unicode string behaves.

I don't think that case-folding on python2 is that reasonable to implement -- at least, not so trivial that it should simply be folded in here.

Reply:
I tried it with ß and saw py2 not doing anything at all with that regardless.  But, someone in `#python` found the example below to vary between `.lower()` and `.upper().lower()` when using Unicode.  The tests don't look to be using Unicode for choice though and the docstring doesn't even mention the expected type for the `choices` parameter so I don't know.  I'm fine with whatever (and it's already merged).  I suppose it could even be argued that my addition of using `str.casefold` at all is an unrelated change.  `:]`

```python2
Python 2.7.15 (default, Dec 11 2018, 00:01:04) 
[GCC 6.3.0 20170516] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> x = u'µıſςϐϑϕϖϰϱϵẛι'
>>> x
u'\xb5\u0131\u017f\u03c2\u03d0\u03d1\u03d5\u03d6\u03f0\u03f1\u03f5\u1e9b\u1fbe'
>>> x.lower()
u'\xb5\u0131\u017f\u03c2\u03d0\u03d1\u03d5\u03d6\u03f0\u03f1\u03f5\u1e9b\u1fbe'
>>> x.upper().lower()
u'\u03bcis\u03c3\u03b2\u03b8\u03c6\u03c0\u03ba\u03c1\u03b5\u1e61\u03b9'
>>> x = 'µıſςϐϑϕϖϰϱϵẛι'
>>> x
'\xc2\xb5\xc4\xb1\xc5\xbf\xcf\x82\xcf\x90\xcf\x91\xcf\x95\xcf\x96\xcf\xb0\xcf\xb1\xcf\xb5\xe1\xba\x9b\xe1\xbe\xbe'
>>> x.lower()
'\xc2\xb5\xc4\xb1\xc5\xbf\xcf\x82\xcf\x90\xcf\x91\xcf\x95\xcf\x96\xcf\xb0\xcf\xb1\xcf\xb5\xe1\xba\x9b\xe1\xbe\xbe'
>>> x.upper().lower()
'\xc2\xb5\xc4\xb1\xc5\xbf\xcf\x82\xcf\x90\xcf\x91\xcf\x95\xcf\x96\xcf\xb0\xcf\xb1\xcf\xb5\xe1\xba\x9b\xe1\xbe\xbe'
```